nmmo/task/base_predicate.py

```diff
@@ -0,0 +1,430 @@
+# TODO: the below line will be gone after implementation
+# pylint: disable=unnecessary-pass
+from typing import List
+
+from nmmo.task.predicate import Predicate
+from nmmo.task.game_state import TeamGameState
+from nmmo.systems import skill as Skill
+from nmmo.lib.material import Material
+from nmmo.lib import utils
+
+from nmmo.entity.entity import EntityState
+
+
+class Timer(Predicate):
+  def __init__(self, num_tick: int):
+    super().__init__(num_tick)
+    self.num_tick = num_tick
+
+  def __call__(self, team_gs: TeamGameState, ent_id):
+    """True if the current tick is within the specified num_tick
+       Otherwise false.
+    """
+    return team_gs.tick <= self.num_tick
+
+
+# CHECK ME: maybe this should be the default task?
+class LiveLong(Timer):
+  def __call__(self, team_gs, ent_id):
+    """True if the health of agent (ent_id) is greater than 0.
+       Otherwise false.
+    """
+    super().__call__(team_gs, ent_id)
+    agent = team_gs.entity_or_none(ent_id)
+    if agent:
+      return agent.time_alive >= self.num_tick
+
+    return False
+
+
+# each agent is rewarded if the alive teammate is greater than min_size
+class TeamSizeGE(Predicate): # greater than or equal to
+  def __init__(self, num_agent: int):
+    super().__init__(num_agent)
+    self.num_agent = num_agent
+
+  def __call__(self, team_gs, ent_id):
+    """True if the number of alive teammates is greater than or equal to min_size.
+       Otherwise false.
+    """
+    super().__call__(team_gs, ent_id)
+    if team_gs.pop_id in team_gs.alive_byteam:
+      return len(team_gs.alive_byteam[team_gs.pop_id]) >= self.num_agent
+
+    return False
+
+
+class ProtectAgent(Predicate):
+  def __init__(self, agents: List[int]):
+    #   is this right?
+    super().__init__(agents)
+    self.agents = agents
+
+  def __call__(self, team_gs, ent_id):
+    """True if the specified agents (list of ent_id) are ALL alive
+       Otherwise false.
+    """
+    super().__call__(team_gs, ent_id)
+
+    # cache the result
+    team_gs.cache_result[self.name] = False
```

Comment:
I don't understand how the caching works, it seems like you only use team_gs.cache_result if self.agents is empty? But what's the point of the cache?

Reply:
Cache will have to be implemented differently, based on the new design. I'll also do caching

Reply:
Caching inside the predicates are all removed and moved to the wrapper.